README.md
# Rack::SslEnforcer [![Build Status](https://travis-ci.org/tobmatth/rack-ssl-enforcer.svg?branch=master)](https://travis-ci.org/tobmatth/rack-ssl-enforcer)

Rack::SslEnforcer is a simple Rack middleware to enforce SSL connections. As of Version 0.2.0, Rack::SslEnforcer marks
Cookies as secure by default (HSTS must be set manually).

Tested against Ruby 1.8.7, 1.9.2, 1.9.3, 2.0.0, 2.1.10, 2.2.7, 2.3.4, 2.4.1, ruby-head, REE and the latest versions of Rubinius & JRuby.

## Installation

The simplest way to install Rack::SslEnforcer is to use [Bundler](http://gembundler.com).

Add Rack::SslEnforcer to your `Gemfile`:

```ruby
 gem 'rack-ssl-enforcer'
```

### Installing on Sinatra / Padrino application

In order for Rack::SslEnforcer to properly work it has to be at the top
of the Rack Middleware.

Using `enable :session` will place Rack::Session::Cookie before Rack::Ssl::Enforcer
and will prevent Rack::Ssl::Enforcer from marking cookies as secure.

To fix this issue do not use `enable :sessions` instead add the
Rack::Session::Cookie middleware after Rack::Ssl::Enforcer.

Eg:

```ruby
use Rack::SslEnforcer
set :session_secret, 'asdfa2342923422f1adc05c837fa234230e3594b93824b00e930ab0fb94b'

#Enable sinatra sessions
use Rack::Session::Cookie, :key => '_rack_session',
                           :path => '/',
                           :expire_after => 2592000, # In seconds
                           :secret => settings.session_secret
```

## Basic Usage

If you don't use Bundler, be sure to require Rack::SslEnforcer manually before actually using the middleware:

```ruby
 require 'rack/ssl-enforcer'
 use Rack::SslEnforcer
```

To use Rack::SslEnforcer in your Rails application, add the following line to your application config file (`config/application.rb` for Rails 3 and above, `config/environment.rb` for Rails 2):

```ruby
config.middleware.insert_before  ActionDispatch::Cookies, Rack::SslEnforcer
```
Cookies should be set before the SslEnforcer processes the headers,
but due to the middleware calls chain Rack::SslEnforcer should be inserted before the ActionDispatch::Cookies.

If all you want is SSL for your whole application, you are done! Otherwise, you can specify some options described below.

## Options

### Host constraints

You can enforce SSL connections only for certain hosts with `:only_hosts`, or prevent certain hosts from being forced to SSL with `:except_hosts`. Constraints can be a `String`, a `Regex` or an array of `String` or `Regex` (possibly mixed), as shown in the following examples:

```ruby
config.middleware.use Rack::SslEnforcer, :only_hosts => 'api.example.com'
# Please note that, for instance, both http://help.example.com/demo and https://help.example.com/demo would be accessible here

config.middleware.use Rack::SslEnforcer, :except_hosts => /[help|blog]\.example\.com$/

config.middleware.use Rack::SslEnforcer, :only_hosts => [/[secure|admin]\.example\.org$/, 'api.example.com']
```

### Path constraints

You can enforce SSL connections only for certain paths with `:only`, prevent certain paths from being forced to SSL with `:except`, or - if you don't care how certain paths are accessed - ignore them with `:ignore`. Constraints can be a `String`, a `Regex` or an array of `String` or `Regex` (possibly mixed), as shown in the following examples:

```ruby
config.middleware.use Rack::SslEnforcer, :only => '/login'
# Please note that, for instance, both http://example.com/demo and https://example.com/demo would be accessible here

config.middleware.use Rack::SslEnforcer, :only => %r{^/admin/}

config.middleware.use Rack::SslEnforcer, :except => ['/demo', %r{^/public/}]

config.middleware.use Rack::SslEnforcer, :ignore => '/assets'

# You can also combine multiple constraints
config.middleware.use Rack::SslEnforcer, :only => '/cart', :ignore => %r{/assets}, :strict => true

# And ignore based on blocks
config.middleware.use Rack::SslEnforcer, :ignore => lambda { |request| request.env["HTTP_X_IGNORE_SSL_ENFORCEMENT"] == "magic" }
```

### Method constraints

You can enforce SSL connections only for certain HTTP methods with `:only_methods`, or prevent certain HTTP methods from being forced to SSL with `:except_methods`. Constraints can be a `String` or an array of `String`, as shown in the following examples:

```ruby
# constraint as a String
config.middleware.use Rack::SslEnforcer, :only_methods => 'POST'
# Please note that, for instance, GET requests would be accessible via SSL and non-SSL connection here

config.middleware.use Rack::SslEnforcer, :except_methods => ['GET', 'HEAD']
```

Note: The `:hosts` constraint takes precedence over the `:path` constraint. Please see the tests for examples.


### Environment constraints

You can enforce SSL connections only for certain environments with `:only_environments` or prevent certain environments from being forced to SSL with `:except_environments`.
Environment constraints may be a `String`, a `Regex` or an array of `String` or `Regex` (possibly mixed), as shown in the following examples:

```ruby
config.middleware.use Rack::SslEnforcer, :except_environments => 'development'

config.middleware.use Rack::SslEnforcer, :except_environments => /^[0-9a-f]+_local$/i

config.middleware.use Rack::SslEnforcer, :only_environments => ['production', /^QA/]
```

Note: The `:environments` constraint requires one the following environment variables to be set: `RACK_ENV`, `RAILS_ENV`, `ENV`.

### User agent constraints

You can enforce SSL connections only for certain user agents with `:only_agents` or prevent certain user agents from being forced to SSL with `:except_agents`.
User agent constraints may be a `String`, a `Regex` or an array of `String` or `Regex` (possibly mixed), as shown in the following examples:

```ruby
config.middleware.use Rack::SslEnforcer, :except_agents => 'Googlebot'

config.middleware.use Rack::SslEnforcer, :except_agents => /[Googlebot|bingbot]/

config.middleware.use Rack::SslEnforcer, :only_agents => ['test-secu-bot', /custom-crawler[0-9a-f]/]
```

### Force-redirection to non-SSL connection if constraint is not matched

Use the `:strict` option to force non-SSL connection for all requests not matching the constraints you set. Examples:

```ruby
config.middleware.use Rack::SslEnforcer, :only => ["/login", /\.xml$/], :strict => true
# https://example.com/demo would be redirected to http://example.com/demo

config.middleware.use Rack::SslEnforcer, :except_hosts => 'demo.example.com', :strict => true
# https://demo.example.com would be redirected to http://demo.example.com
```

### Automatic method constraints

In the case where you have matching URLs with different HTTP methods – for instance Rails RESTful routes: `GET /users`, `POST /users`, `GET /user/:id` and `PUT /user/:id` – you may need to force POST and PUT requests to SSL connection but redirect to non-SSL connection on GET.

```ruby
config.middleware.use Rack::SslEnforcer, :only => [%r{^/users/}], :mixed => true
```

The above will allow you to POST/PUT from the secure/non-secure URLs keeping the original schema.

### HTTP Strict Transport Security (HSTS)

To set HSTS expiry and subdomain inclusion (defaults respectively to `one year` and `true`).

```ruby
config.middleware.use Rack::SslEnforcer, :hsts => { :expires => 500, :subdomains => false, :preload => false }
config.middleware.use Rack::SslEnforcer, :hsts => true # equivalent to { :expires => 31536000, :subdomains => true, :preload => false }
```
Please note that the strict option disables HSTS.

### Redirect to specific URL (e.g. if you're using a proxy)

You might need the `:redirect_to` option if the requested URL can't be determined.

```ruby
config.middleware.use Rack::SslEnforcer, :redirect_to => 'https://example.org'
```

### Redirect with specific HTTP status code

By default it redirects with HTTP status code 301(Moved Permanently). Sometimes you might need to redirect with different HTTP status code:

```ruby
config.middleware.use Rack::SslEnforcer, :redirect_code => 302
```

### Custom HTTP port

If you're using a different port than the default (80) for HTTP, you can specify it with the `:http_port` option:

```ruby
config.middleware.use Rack::SslEnforcer, :http_port => 8080
```

### Custom HTTPS port

If you're using a different port than the default (443) for HTTPS, you can specify it with the `:https_port` option:

```ruby
config.middleware.use Rack::SslEnforcer, :https_port => 444
```

### Secure cookies disabling

Finally you might want to share a cookie based session between HTTP and HTTPS.
This is not possible by default with Rack::SslEnforcer for [security reasons](http://en.wikipedia.org/wiki/HTTP_cookie#Cookie_theft_and_session_hijacking).

Nevertheless, you can set the `:force_secure_cookies` option to `false` in order to be able to share a cookie based session between HTTP and HTTPS:

```ruby
config.middleware.use Rack::SslEnforcer, :only => "/login", :force_secure_cookies => false
```

But be aware that if you do so, you have to make sure that the content of you cookie is encoded.
This can be done using a coder with [Rack::Session::Cookie](https://github.com/rack/rack/blob/master/lib/rack/session/cookie.rb#L28-42).

### Running code before redirect

You may want to run some code before rack-ssl-enforcer forces a redirect.  This code could do something like log that a redirect was done, or if this is used in a Rails app, keeping the flash when redirecting:


```ruby
config.middleware.use Rack::SslEnforcer, :only => '/login', :before_redirect => Proc.new { |request|
  #keep flash on redirect
  request.session[:flash].keep if !request.session.nil? && request.session.key?('flash') && !request.session['flash'].empty?
}
```

### Custom or empty response body

The default response body is:

```html
<html><body>You are being <a href="https://www.example.org/">redirected</a>.</body></html>
```

To supply a custom message, provide a string as `:redirect_html`:

```ruby
config.middleware.use Rack::SslEnforcer, :redirect_html => 'Redirecting!'
```

To supply a series of responses for Rack to chunk, provide a [permissable Rack body object](http://rack.rubyforge.org/doc/SPEC.html):

```ruby
config.middleware.use Rack::SslEnforcer, :redirect_html => ['<html>','<body>','Hello!','</body>','</html>']
```

To supply an empty response body, provide :redirect_html as false:

```ruby
config.middleware.use Rack::SslEnforcer, :redirect_html => false
```

## Deployment

If you run your application behind a proxy (e.g. Nginx) you may need to do some configuration on that side. If you don't you may experience an infinite redirect loop.

The reason this happens is that Rack::SslEnforcer can't detect if you are running SSL or not. The solution is to have your front-end server send extra headers for Rack::SslEnforcer to identify the request protocol.

### Nginx

In the `location` block for your app's SSL configuration, include the following proxy header configuration:

`proxy_set_header   X-Forwarded-Proto https;`

### Nginx behind Load Balancer

The following instruction has been tested behind AWS ELB, but can be adapted to work behind any load balancer.
Specifically, the options below account for termination of SSL at the load balancer (HTTP only communication between load balancer and server), and HTTP only health checks.

In the `location` block for your app's SSL configuration, reference the following proxy header configurations:

```
proxy_set_header X-Forwarded-Proto $http_x_forwarded_proto;
proxy_set_header Host $http_host;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for
proxy_redirect off;
```

In `config/application.rb` for Rails 3 and above, `config/environment.rb` for Rails 2, work off the following line:

```ruby
config.middleware.use Rack::SslEnforcer, ignore: lambda { |request| request.env["HTTP_X_FORWARDED_PROTO"].blank? }, strict: true
```

This ignores ELB healthchecks and development environment as ELB healthchecks aren't forwarded requests, so it wouldn't have the forwarded proto header.

Same goes for when running without a proxy (like developemnt locally), making `:except_environments => 'development'` unnecessary.

In Sinatra or other non-rails ruby applications, work off the following line:
```ruby
use Rack::SslEnforcer, :except_environments => ['development', 'test'], :ignore => '/healthcheck'
```

### Passenger

Or, if you're using mod_rails/passenger (which will ignore the proxy_xxx directives):

`passenger_set_cgi_param HTTP_X_FORWARDED_PROTO https;`

If you're sharing a single `server` block for http AND https access you can add:

`passenger_set_cgi_param HTTP_X_FORWARDED_PROTO $scheme;`

This makes sure that Rack::SslEnforcer knows it's being accessed over SSL. Just restart Nginx for these changes to take effect.

## TODO

* Cleanup tests

#### Contributors

[https://github.com/tobmatth/rack-ssl-enforcer/graphs/contributors](https://github.com/tobmatth/rack-ssl-enforcer/graphs/contributors)

## Credits

Flagging cookies as secure functionality and HSTS support is greatly inspired by [Joshua Peek's Rack::SSL](https://github.com/josh/rack-ssl).

## Note on Patches / Pull Requests

* Fork the project.
* Code your feature addition or bug fix.
* **Add tests for it.** This is important so we don't break it in a future version unintentionally.
* Commit, do not mess with Rakefile or version number. If you want to have your own version, that's fine but bump version in a commit by itself so we can ignore it when merging.
* Send a pull request. Bonus points for topic branches.

## Copyright

Copyright (c) 2010-2013 Tobias Matthies. See [LICENSE](https://github.com/tobmatth/rack-ssl-enforcer/blob/master/LICENSE) for details.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/tobmatth/rack-ssl-enforcer/tree/HEAD)

[Full Changelog](https://github.com/tobmatth/rack-ssl-enforcer/compare/v0.2.9...HEAD)

**Closed issues:**

- HSTS Implementation [\#92](https://github.com/tobmatth/rack-ssl-enforcer/issues/92)
- Does not set session cookie as secure [\#91](https://github.com/tobmatth/rack-ssl-enforcer/issues/91)
- HSTS and secure cookies w/o redirect? [\#89](https://github.com/tobmatth/rack-ssl-enforcer/issues/89)
- ERR\_CONNECTION\_REFUSED [\#85](https://github.com/tobmatth/rack-ssl-enforcer/issues/85)
- Issue with IE only... strict true not working [\#80](https://github.com/tobmatth/rack-ssl-enforcer/issues/80)

**Merged pull requests:**

- README: Typo [\#100](https://github.com/tobmatth/rack-ssl-enforcer/pull/100) ([olleolleolle](https://github.com/olleolleolle))
- Travis: jruby-9.1.13.0 [\#99](https://github.com/tobmatth/rack-ssl-enforcer/pull/99) ([olleolleolle](https://github.com/olleolleolle))
- Travis: Use jruby-9.1.10.0 in CI matrix [\#97](https://github.com/tobmatth/rack-ssl-enforcer/pull/97) ([olleolleolle](https://github.com/olleolleolle))
- update instructions for configuring nginx behind ELB, add for sinatra [\#96](https://github.com/tobmatth/rack-ssl-enforcer/pull/96) ([bmishkin](https://github.com/bmishkin))
- Travis: use JRuby 9.1.7.0 [\#94](https://github.com/tobmatth/rack-ssl-enforcer/pull/94) ([olleolleolle](https://github.com/olleolleolle))
- README: Use shiny SVG badge [\#93](https://github.com/tobmatth/rack-ssl-enforcer/pull/93) ([olleolleolle](https://github.com/olleolleolle))
- Specify to insert SslEnforcer before the Cookies middleware in Rails [\#90](https://github.com/tobmatth/rack-ssl-enforcer/pull/90) ([DimaSamodurov](https://github.com/DimaSamodurov))
- Add example for HSTS preload option on README [\#87](https://github.com/tobmatth/rack-ssl-enforcer/pull/87) ([camelmasa](https://github.com/camelmasa))
- Adding MIT license to the gemspec. [\#86](https://github.com/tobmatth/rack-ssl-enforcer/pull/86) ([reiz](https://github.com/reiz))

## [v0.2.9](https://github.com/tobmatth/rack-ssl-enforcer/tree/v0.2.9) (2015-07-22)
[Full Changelog](https://github.com/tobmatth/rack-ssl-enforcer/compare/v0.2.8...v0.2.9)

**Closed issues:**

- Infinite redirects behind AWS ELB [\#82](https://github.com/tobmatth/rack-ssl-enforcer/issues/82)
- Issue with Redirects [\#81](https://github.com/tobmatth/rack-ssl-enforcer/issues/81)
- POST requests [\#79](https://github.com/tobmatth/rack-ssl-enforcer/issues/79)
- How to handle URI::InvalidURIError? [\#78](https://github.com/tobmatth/rack-ssl-enforcer/issues/78)
- Cookie session state shared across http and https without disabling force\_secure\_cookies [\#58](https://github.com/tobmatth/rack-ssl-enforcer/issues/58)
- :strict option + AJAX requests [\#36](https://github.com/tobmatth/rack-ssl-enforcer/issues/36)

**Merged pull requests:**

- Add HSTS preload option [\#84](https://github.com/tobmatth/rack-ssl-enforcer/pull/84) ([gorism](https://github.com/gorism))
- added Nginx behind Load Balancer section to readme [\#83](https://github.com/tobmatth/rack-ssl-enforcer/pull/83) ([gnitnuj](https://github.com/gnitnuj))
- respect rack.url\_scheme header for proxied SSL when HTTP\_X\_FORWARDED\_PROTO blank [\#77](https://github.com/tobmatth/rack-ssl-enforcer/pull/77) ([grantspeelman](https://github.com/grantspeelman))

## [v0.2.8](https://github.com/tobmatth/rack-ssl-enforcer/tree/v0.2.8) (2014-07-18)
[Full Changelog](https://github.com/tobmatth/rack-ssl-enforcer/compare/v0.2.7...v0.2.8)

**Closed issues:**

- Already encoded url parameters get encoded again when redirecting [\#75](https://github.com/tobmatth/rack-ssl-enforcer/issues/75)
- Release new version! \<3 [\#73](https://github.com/tobmatth/rack-ssl-enforcer/issues/73)

**Merged pull requests:**

- Do not encode already encoded url parameters when redirecting. [\#76](https://github.com/tobmatth/rack-ssl-enforcer/pull/76) ([oveddan](https://github.com/oveddan))
- Enable ignore blocks [\#74](https://github.com/tobmatth/rack-ssl-enforcer/pull/74) ([danielevans](https://github.com/danielevans))

## [v0.2.7](https://github.com/tobmatth/rack-ssl-enforcer/tree/v0.2.7) (2014-05-23)
[Full Changelog](https://github.com/tobmatth/rack-ssl-enforcer/compare/v0.2.6...v0.2.7)

**Fixed bugs:**

- Vertical pipe characters in the URL cause an URI::InvalidURIError [\#47](https://github.com/tobmatth/rack-ssl-enforcer/issues/47)

**Closed issues:**

- Support for ruby 2.0 and Rails 4 [\#72](https://github.com/tobmatth/rack-ssl-enforcer/issues/72)
- Running code before redirect not working [\#70](https://github.com/tobmatth/rack-ssl-enforcer/issues/70)
- combine strict and non strict behaviour [\#69](https://github.com/tobmatth/rack-ssl-enforcer/issues/69)
- Is there a way to combine mutiple only, multiple ignore with strict [\#68](https://github.com/tobmatth/rack-ssl-enforcer/issues/68)
- Enforcing won't preserve HTTP methods [\#65](https://github.com/tobmatth/rack-ssl-enforcer/issues/65)
- Rack::SslEnforcer options mess up with 'localhost' [\#64](https://github.com/tobmatth/rack-ssl-enforcer/issues/64)
- New rubygems release? [\#60](https://github.com/tobmatth/rack-ssl-enforcer/issues/60)

**Merged pull requests:**

- Fixing issue \#70 - Running code before redirect not working [\#71](https://github.com/tobmatth/rack-ssl-enforcer/pull/71) ([abhasg](https://github.com/abhasg))
- URI encode before passing to URI object to deal with pathological URIs [\#67](https://github.com/tobmatth/rack-ssl-enforcer/pull/67) ([tilthouse](https://github.com/tilthouse))
- Add Ruby 2.1.0 to .travis.yml [\#66](https://github.com/tobmatth/rack-ssl-enforcer/pull/66) ([salimane](https://github.com/salimane))
- Allow for custom, default, or no body when redirecting [\#61](https://github.com/tobmatth/rack-ssl-enforcer/pull/61) ([kcm](https://github.com/kcm))

## [v0.2.6](https://github.com/tobmatth/rack-ssl-enforcer/tree/v0.2.6) (2013-09-18)

test/helper.rb
require 'rubygems'
require 'test/unit'
require 'shoulda'
require 'rack/mock'
require 'rack/test'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rack/ssl-enforcer'

class Test::Unit::TestCase
  include Rack::Test::Methods

  def app; Rack::Lint.new(@app); end

  def mock_app(options_or_options_array = {})
    main_app = lambda { |env|
      request = Rack::Request.new(env)
      headers = {'Content-Type' => "text/html"}
      headers['Set-Cookie'] = "id=1; path=/\ntoken=abc; path=/; secure; HttpOnly"
      [200, headers, ['Hello world!']]
    }

    builder = Rack::Builder.new
    options_or_options_array = [options_or_options_array] unless options_or_options_array.is_a?(Array)
    Array(options_or_options_array).each do |options|
      builder.use Rack::SslEnforcer, options
    end
    builder.run main_app
    @app = builder.to_app
  end

end


test/test_rack-ssl-enforcer.rb
require 'helper'

class TestRackSslEnforcer < Test::Unit::TestCase
  
  context 'that has no :redirect_to set' do
    setup { mock_app }
    
    should 'respond with a ssl redirect to plain-text requests' do
      get 'http://www.example.org/'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/', last_response.location
    end
    
    should 'respond with a ssl redirect to plain-text requests and keep params' do
      get 'http://www.example.org/admin?token=33'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/admin?token=33', last_response.location
    end
    
    #heroku / etc do proxied SSL
    #http://github.com/pivotal/refraction/issues/issue/2
    should 'respect X-Forwarded-Proto header for proxied SSL' do
      get 'http://www.example.org/', {}, { 'HTTP_X_FORWARDED_PROTO' => 'http', 'rack.url_scheme' => 'http' }
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/', last_response.location
    end
    
    should 'respond not redirect ssl requests' do
      get 'https://www.example.org/'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
    
    should 'respond not redirect ssl requests and respect X-Forwarded-Proto header for proxied SSL' do
      get 'http://www.example.org/', {}, { 'HTTP_X_FORWARDED_PROTO' => 'https', 'rack.url_scheme' => 'http' }
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
    
    should 'use default https port when redirecting non-standard http port to ssl' do
      get 'http://example.org:81/', {}, { 'rack.url_scheme' => 'http' }
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/', last_response.location
    end
  end
  
  context 'that has :redirect_to set' do
    setup { mock_app :redirect_to => 'https://www.google.com' }
    
    should 'respond with a ssl redirect to plain-text requests and redirect to :redirect_to' do
      get 'http://www.example.org/'
      assert_equal 301, last_response.status
      assert_equal 'https://www.google.com', last_response.location
    end
    
    should 'respond not redirect ssl requests' do
      get 'https://www.example.org/'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
  end
  
  context 'that has regex pattern as only option' do
    setup { mock_app :only => /^\/admin/ }
    
    should 'respond with a ssl redirect for /admin path' do
      get 'http://www.example.org/admin'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/admin', last_response.location
    end
    
    should 'respond not redirect ssl requests' do
      get 'http://www.example.org/foo'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
  end
  
  context 'that has path as only option' do
    setup { mock_app :only => "/login" }
    
    should 'respond with a ssl redirect for /login path' do
      get 'http://www.example.org/login'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/login', last_response.location
    end
    
    should 'respond not redirect ssl requests' do
      get 'http://www.example.org/foo/'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
  end
  
  context 'that has array of regex pattern & path as only option' do
    setup { mock_app :only => [/\.xml$/, "/login"] }
    
    should 'respond with a ssl redirect for /login path' do
      get 'http://www.example.org/login'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/login', last_response.location
    end
    
    should 'respond with a ssl redirect for /admin path' do
      get 'http://www.example.org/users.xml'
      assert_equal 301, last_response.status
      assert_equal 'https://www.example.org/users.xml', last_response.location
    end
    
    should 'respond not redirect ssl requests' do
      get 'http://www.example.org/foo/'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
  end
  
  context 'that has array of regex pattern & path as only option with strict option' do
    setup { mock_app :only => [/\.xml$/, "/login"], :strict => true }
    
    should 'respond with a http redirect from non-allowed https url' do
      get 'https://www.example.org/foo/'
      assert_equal 301, last_response.status
      assert_equal 'http://www.example.org/foo/', last_response.location
    end
    
    should 'respond from allowed https url' do
      get 'https://www.example.org/login'
      assert_equal 200, last_response.status
      assert_equal 'Hello world!', last_response.body
    end
    
    should 'use default https port when redirecting non-standard ssl port to http' do
      get 'https://example.org:81/', {}, { 'rack.url_scheme' => 'https' }
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/', last_response.location
    end
  end
  
end

lib/rack/ssl-enforcer/constraint.rb
class SslEnforcerConstraint
  def initialize(name, rule, request)
    @name    = name
    @rule    = rule
    @request = request
  end

  def matches?
    if @rule.is_a?(String) && [:only, :except].include?(@name)
      result = tested_string[0, @rule.size].send(operator, @rule)
    elsif @rule.respond_to?(:call)
      result = @rule.call(@request)
    else
      result = tested_string.send(operator, @rule)
    end

    negate_result? ? !result : result
  end

private

  def negate_result?
    @name.to_s =~ /except/
  end

  def operator
    @rule.is_a?(Regexp) ? "=~" : "=="
  end

  def tested_string
    case @name.to_s
    when /hosts/
      @request.host
    when /methods/
      @request.request_method
    when /environments/
      ENV["RACK_ENV"] || ENV["RAILS_ENV"] || ENV["ENV"]
    when /agents/
      @request.user_agent
    else
      @request.path
    end
  end
end


lib/rack/ssl-enforcer.rb
require 'rack/ssl-enforcer/constraint'

module Rack

  class SslEnforcer

    CONSTRAINTS_BY_TYPE = {
      :hosts        => [:only_hosts, :except_hosts],
      :agents       => [:only_agents, :except_agents],
      :path         => [:only, :except],
      :methods      => [:only_methods, :except_methods],
      :environments => [:only_environments, :except_environments]
    }

    # Warning: If you set the option force_secure_cookies to false, make sure that your cookies
    # are encoded and that you understand the consequences (see documentation)
    def initialize(app, options={})
      default_options = {
        :redirect_to          => nil,
        :redirect_code        => nil,
        :strict               => false,
        :mixed                => false,
        :hsts                 => nil,
        :http_port            => nil,
        :https_port           => nil,
        :force_secure_cookies => true,
        :redirect_html        => nil,
        :before_redirect      => nil
      }
      CONSTRAINTS_BY_TYPE.values.each do |constraints|
        constraints.each { |constraint| default_options[constraint] = nil }
      end

      @app, @options = app, default_options.merge(options)
    end

    def call(env)
      @request = Rack::Request.new(env)

      return @app.call(env) if ignore?

      @scheme = if enforce_ssl?
        'https'
      elsif enforce_non_ssl?
        'http'
      end

      if redirect_required?
        call_before_redirect
        modify_location_and_redirect
      elsif ssl_request?
        status, headers, body = @app.call(env)
        flag_cookies_as_secure!(headers) if @options[:force_secure_cookies]
        set_hsts_headers!(headers) if @options[:hsts] && !@options[:strict]
        [status, headers, body]
      else
        @app.call(env)
      end
    end

  private

    def redirect_required?
      scheme_mismatch? || host_mismatch?
    end

    def ignore?
      if @options[:ignore]
        rules = [@options[:ignore]].flatten.compact
        rules.any? do |rule|
          SslEnforcerConstraint.new(:ignore, rule, @request).matches?
        end
      else
        false
      end
    end

    def scheme_mismatch?
      @scheme && @scheme != current_scheme
    end

    def host_mismatch?
      destination_host && destination_host != @request.host
    end

    def call_before_redirect
       @options[:before_redirect].call(@request) unless @options[:before_redirect].nil?
    end

    def modify_location_and_redirect
      location = "#{current_scheme}://#{@request.host}#{@request.fullpath}"
      location = replace_scheme(location, @scheme)
      location = replace_host(location, @options[:redirect_to])
      redirect_to(location)
    rescue URI::InvalidURIError
      [400, { 'Content-Type' => 'text/plain'}, []]
    end

    def redirect_to(location)
      body = []
      body << "<html><body>You are being <a href=\"#{location}\">redirected</a>.</body></html>" if @options[:redirect_html].nil?
      body << @options[:redirect_html] if @options[:redirect_html].is_a?(String)
      body = @options[:redirect_html] if @options[:redirect_html].respond_to?('each')

      headers = { 'Content-Type' => 'text/html', 'Location' => location }
      set_hsts_headers!(headers) if @options[:hsts] && !@options[:strict]

      [@options[:redirect_code] || 301, headers, body]
    end

    def ssl_request?
      current_scheme == 'https'
    end

    def destination_host
      if @options[:redirect_to]
        host_parts = URI.split(@options[:redirect_to])
        host_parts[2] || host_parts[5]
      end
    end

    # Fixed in rack >= 1.3
    def current_scheme
      if @request.env['HTTPS'] == 'on' || @request.env['HTTP_X_SSL_REQUEST'] == 'on'
        'https'
      elsif @request.env['HTTP_X_FORWARDED_PROTO']
        @request.env['HTTP_X_FORWARDED_PROTO'].split(',')[0] || @request.scheme
      else
        @request.scheme
      end
    end

    def enforce_ssl_for?(keys)
      provided_keys = keys.select { |key| @options[key] }
      if provided_keys.empty?
        true
      else
        provided_keys.all? do |key|
          rules = [@options[key]].flatten.compact
          rules.send([:except_hosts, :except_agents, :except_environments, :except].include?(key) ? :all? : :any?) do |rule|
            SslEnforcerConstraint.new(key, rule, @request).matches?
          end
        end
      end
    end

    def enforce_non_ssl?
      @options[:strict] || @options[:mixed] && !(@request.request_method == 'PUT' || @request.request_method == 'POST')
    end

    def enforce_ssl?
      CONSTRAINTS_BY_TYPE.inject(true) do |memo, (type, keys)|
        memo && enforce_ssl_for?(keys)
      end
    end

    def replace_scheme(uri, scheme)
      return uri if not scheme_mismatch?

      port = adjust_port_to(scheme)
      uri_parts = URI.split(uri)
      uri_parts[3] = port unless port.nil?
      uri_parts[0] = scheme
      URI::HTTP.new(*uri_parts).to_s
    end

    def replace_host(uri, host)
      return uri unless host_mismatch?

      host_parts = URI.split(host)
      new_host = host_parts[2] || host_parts[5]
      uri_parts = URI.split(uri)
      uri_parts[2] = new_host
      URI::HTTPS.new(*uri_parts).to_s
    end

    def adjust_port_to(scheme)
      if scheme == 'https'
        @options[:https_port] if @options[:https_port] && @options[:https_port] != URI::HTTPS.default_port
      elsif scheme == 'http'
        @options[:http_port] if @options[:http_port] && @options[:http_port] != URI::HTTP.default_port
      end
    end

    # see http://en.wikipedia.org/wiki/HTTP_cookie#Cookie_theft_and_session_hijacking
    def flag_cookies_as_secure!(headers)
      if cookies = headers['Set-Cookie']
        # Support Rails 2.3 / Rack 1.1 arrays as headers
        unless cookies.is_a?(Array)
          cookies = cookies.split("\n")
        end

        headers['Set-Cookie'] = cookies.map do |cookie|
          cookie !~ /(^|;\s)secure($|;)/ ? "#{cookie}; secure" : cookie
        end.join("\n")
      end
    end

    # see http://en.wikipedia.org/wiki/Strict_Transport_Security
    def set_hsts_headers!(headers)
      opts = { :expires => 31536000, :subdomains => true, :preload => false }
      opts.merge!(@options[:hsts]) if @options[:hsts].is_a? Hash
      value  = "max-age=#{opts[:expires]}"
      value += "; includeSubDomains" if opts[:subdomains]
      value += "; preload" if opts[:preload]
      headers.merge!({ 'Strict-Transport-Security' => value })
    end

  end
end
